cli/listener/src/main.rs

```diff
@@ -210,74 +215,417 @@ pub fn get_apy_for_period(
+            ) {
+                ListenerResult::ErrSnapshot(err) => {
+                    println!("Error while obtaining on-chain state.");
+                    err.print_pretty();
+                    self.metrics.errors += 1;
+                    self.get_sleep_time_after_error()
+                }
+                ListenerResult::OkListener(exchange_rate, interval_prices_option) => {
+                    println!(
+                        "Got exchange rate: {}/{}: {} at slot {} and epoch {}.",
+                        exchange_rate.price_lamports_numerator,
+                        exchange_rate.price_lamports_denominator,
+                        exchange_rate.price_lamports_numerator as f32
+                            / exchange_rate.price_lamports_denominator as f32,
+                        exchange_rate.slot,
+                        exchange_rate.epoch,
+                    );
+
+                    match interval_prices_option {
+                        None => println!(
+                            "No interval price could be produced, awaiting more data points"
+                        ),
+                        Some(interval_prices) => {
+                            println!("{}", interval_prices);
+                            self.metrics.solido_average_interval_price = Some(interval_prices);
+                        }
+                    }
+                    self.get_sleep_time()
+                }
+                ListenerResult::ErrListener(err) => {
+                    println!("Error in listener.");
+                    err.print_pretty();
+                    self.get_sleep_time_after_error()
+                }
+            };
+            // Update metrics snapshot.
+            *self.metrics_snapshot.lock().unwrap() = Arc::new(self.metrics.clone());
+            std::thread::sleep(sleep_time);
+        }
+    }
+
+    fn get_sleep_time_after_error(&mut self) -> Duration {
+        // For the sleep time we use exponential backoff with jitter [1]. By taking
+        // the time since the last success as the target sleep time, we get
+        // exponential backoff. We clamp this to ensure we don't wait indefinitely.
+        // 1: https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+        let time_since_last_success = self.last_read_success.elapsed();
+        let min_sleep_time = Duration::from_secs_f32(0.2);
+        let max_sleep_time = Duration::from_secs_f32(300.0);
+        let target_sleep_time = time_since_last_success.clamp(min_sleep_time, max_sleep_time);
+        let sleep_time = self
+            .rng
+            .gen_range(Duration::from_secs(0)..target_sleep_time);
+        println!("Sleeping {:?} after error ...", sleep_time);
+        sleep_time
+    }
+
+    pub fn get_sleep_time(&self) -> Duration {
+        // Sleep until is time to get the next exchange rate.
+        let sleep_time = std::time::Duration::from_secs(self.opts.poll_frequency_seconds as u64);
```

Comment:
```suggestion
        let sleep_time = std::time::Duration::from_secs(self.opts.poll_interval_seconds as u64);
```

Technically it’s the poll interval, not the poll frequency. Frequency has units of time<sup>-1</sup>.